### DISCUSSION TITLE
Constrain nodes that are before/after a known time

### DISCUSSION THREAD

**Discussion by jmintz4** [google/or-tools#3563] [Routing (and legacy CP) questions] [answered]
In my routing problem, I am enforcing a minimum break time of 10 hours that must cross over midnight.  Currently, I am using dummy nodes at midnight, which I know is not ideal.  See the code below.   

```py
dummy_node = 4
num_nodes = len(data['time_matrix'])
break_time = 600     # minutes

break_aa = sum([(routing.NextVar(manager.NodeToIndex(j)) == dummy_node) \
                * (time_dimension.CumulVar(manager.NodeToIndex(dummy_node)) \
                   - time_dimension.CumulVar(manager.NodeToIndex(j))) for j in range(num_nodes)])
break_bb = sum([(routing.NextVar(manager.NodeToIndex(dummy_node)) == j) \
                * (time_dimension.CumulVar(manager.NodeToIndex(j)) \
                   - time_dimension.CumulVar(manager.NodeToIndex(dummy_node))) for j in range(num_nodes)])
routing.solver().Add((break_aa + break_bb ) >= break_time)
```

Of course I know the minute of midnight because the dummy_node is constrained there.
So instead of using the dummy node, how could I code this to grab the two closest nodes before and after midnight to constrain the sum of their abs(CumulVar - midnight)?

Note, I am already ensuring that no travel time crosses over midnight.

**Comment by jmintz4** [comment]
This is what I have so far, but it is returning an error set.
For each vehicle, take the nodes before and after the 35 minute mark, and ensure the difference of their CumulVar is >= 10.
Any ideas why this would result in an error set?

built-in function RoutingIndexManager_NodeToIndex returned a result with an error set

```py
    for veh in range(data['num_vehicles']):
        diff = sum([(routing.VehicleVar(manager.NodeToIndex(j)) == veh) \
                    * (time_dimension.CumulVar(manager.NodeToIndex(j)) < 35) \
                    * (time_dimension.CumulVar(manager.NodeToIndex(routing.NextVar(manager.NodeToIndex(j)))) > 35) \
                    * ((time_dimension.CumulVar(manager.NodeToIndex(routing.NextVar(manager.NodeToIndex(j))))) \
                        - (time_dimension.CumulVar(manager.NodeToIndex(j)))) \
                    for j in range(num_nodes)])
        routing.solver().Add(diff >= 10)
```

**Reply by Mizux** [reply]
You can't have `manager.NodeToIndex(routing.NextVar(...))` because `NextVar()` return an object `IntVar` while `NodeToIndex()` want an integer (i.e. a node index) as parameter

**Comment by jmintz4** [comment]
Thanks for the reply.  My next attempt at constraining the nodes on either side of CumulVar=35 is the following. 
However, this does not return a solution, even with a very relaxed constraint of routing.solver().Add(diff >= 2)
Do you see anything wrong with this code?
```py
for ii in range(1, num_nodes):
    diff = sum([(routing.NextVar(manager.NodeToIndex(ii)) == jj) \
                * (time_dimension.CumulVar(manager.NodeToIndex(ii)) < 35) \
                * (time_dimension.CumulVar(manager.NodeToIndex(jj)) > 35) \
                * ((time_dimension.CumulVar(manager.NodeToIndex(jj))) \
                   - (time_dimension.CumulVar(manager.NodeToIndex(ii)))) \
                for jj in range(1, num_nodes)])
    if diff != 0:
        routing.solver().Add(diff >= 2)
```

**Reply by Mizux** [reply]
You can test IntExpr here:
```py
if diff != 0:
```
will always be evaluate to zero by the python interpreter IIRC.

You have a "chicken-egg problem".
Your `sum` aka `diff` is a "constraint expression" thus its value is only know when solver evaluate a  solution candidate.
So you can't use its value at model construction (in an python `if` statement) i.e. you want the solution of the problem to define your problem...

You may be interested by `routing.solver().ConditionalExpression()`:
https://github.com/google/or-tools/blob/88645c398f7506969b419b3a02d40fd4fc35d9f6/ortools/constraint_solver/constraint_solver.h#L1278-L1281
https://github.com/google/or-tools/blob/82750ac12f1ee5354e1c7869894d9af3508778f2/ortools/constraint_solver/python/constraint_solver.i#L770

note: `IntVar op IntVar` => `IntExpr`, `IntExpr::Var() => intVar`
ref: https://github.com/google/or-tools/blob/88645c398f7506969b419b3a02d40fd4fc35d9f6/ortools/constraint_solver/constraint_solver.h#L3986-L3987

Advanced topic:  the python `sum()` is working because `IntExpr` overload `__add__()` method
https://github.com/google/or-tools/blob/82750ac12f1ee5354e1c7869894d9af3508778f2/ortools/constraint_solver/python/constraint_solver.i#L347-L360

**Comment by jmintz4** [comment, marked as answer]
Thank you.  I will look into routing.solver().ConditionalExpression()
